tfhe/src/core_crypto/gpu/vec.rs

```diff
@@ -68,69 +111,133 @@ impl<T: Numeric> CudaVec<T> {
+    ///   not be dropped until streams is synchronised
+    pub unsafe fn memset_multi_gpu_async(&mut self, value: T, streams: &CudaStreams)
+    where
+        T: Into<u64>,
+    {
+        for &gpu_index in self.gpu_indexes.clone().iter() {
+            let size = self.len(gpu_index) * std::mem::size_of::<T>();
+            // We check that self is not empty to avoid invalid pointers
+            if size > 0 {
+                cuda_memset_async(
+                    self.as_mut_c_ptr(gpu_index),
+                    value.into(),
+                    size as u64,
+                    streams.ptr[gpu_index as usize],
+                    streams.gpu_indexes[gpu_index as usize],
+                );
+            }
+        }
+    }
+
     /// Copies data from slice into `CudaVec`
     ///
     /// # Safety
     ///
     /// - [CudaStreams::synchronize] __must__ be called after the copy
     /// as soon as synchronization is required
-    pub unsafe fn copy_from_cpu_async(&mut self, src: &[T], streams: &CudaStreams)
+    pub unsafe fn copy_from_cpu_async(&mut self, src: &[T], streams: &CudaStreams, gpu_index: u32)
     where
         T: Numeric,
     {
-        assert!(self.len() >= src.len());
+        assert!(self.len(gpu_index) >= src.len());
         let size = std::mem::size_of_val(src);
 
         // We have to check that src is not empty, because Rust slice with size 0 results in an
         // invalid pointer being passed to copy_to_gpu_async
         if size > 0 {
             cuda_memcpy_async_to_gpu(
-                self.as_mut_c_ptr(),
+                self.as_mut_c_ptr(gpu_index),
                 src.as_ptr().cast(),
                 size as u64,
-                streams.ptr[0],
-                streams.gpu_indexes[0],
+                streams.ptr[gpu_index as usize],
+                streams.gpu_indexes[gpu_index as usize],
             );
         }
     }
 
+    /// Copies data from slice into `CudaVec`
+    ///
+    /// # Safety
+    ///
+    /// - [CudaStreams::synchronize] __must__ be called after the copy
+    /// as soon as synchronization is required
+    pub unsafe fn copy_from_cpu_multi_gpu_async(&mut self, src: &[T], streams: &CudaStreams)
+    where
+        T: Numeric,
+    {
+        self.gpu_indexes.par_iter().for_each(|&gpu_index| {
+            assert!(self.len(gpu_index) >= src.len());
+            let size = std::mem::size_of_val(src);
+
+            // We have to check that src is not empty, because Rust slice with size 0 results in an
+            // invalid pointer being passed to copy_to_gpu_async
+            if size > 0 {
+                cuda_memcpy_async_to_gpu(
+                    self.get_mut_c_ptr(gpu_index),
+                    src.as_ptr().cast(),
+                    size as u64,
+                    streams.ptr[gpu_index as usize],
+                    streams.gpu_indexes[gpu_index as usize],
+                );
+            }
+        });
+    }
```

Comment:
Since this function only calls memset_async, doing a par_iter seems overkill as the actual setting of values  is not done in the memset_call.

Doing par_iter might only be interesting in the non async version

Reply:
No actually I need the par iter for the calls on the GPU to happen asynchronously, even if I'm using different cuda streams. On the C++ side we handle this with open mp.

Reply:
So you mean that I I have 4 streams, 4 operations and 1 thread, I push asynchronously each operation to each stream and then I synchronize the streams but this time in multithreaded, 1 thread per stream, the GPU computations are not going to happen  concurently ?

Reply:
No if we launch in multithreaded 1 operation on 1 stream on each thread the execution will happen asynchronously. But if we don't use different CPU threads for the different streams execution will remain sequential.

Reply:
At least on the C++ side that's what we observed, I haven't checked again here.

Reply:
Yes, but my question was about 1 CPU threads but multiple streams, which should be the setup here, unless ` streams.ptr[gpu_index as usize]` may be the same stream

Reply:
Well with 1 CPU thread what we observed is that even using different cudaStreams we were getting a synchronous execution (cc @pdroalves maybe you can confirm, this dates back a bit for me).

Reply:
The overhead for CUDA calls is far from negligible and it is not async even though the kernel (or memset/memcpy) operation itself will be async of course, and can overlap. Control will only return to the CPU when the kernel's launched.

If you launch two quick asynchronous (like, finish in a few ms) kernels using a single CPU thread, when the second is launched the first will be done and computation will be sequential in practice.  The same happens to memcpy and memset. 

If we launch everything using the same CPU thread we will have a timeline like this:

```
Launch overhead 1 | Execution 1
                               | Launch overhead 2 | Execution 2
                                                                | Launch overhead 3 | Execution 3
...
```

However, if you run each in a different CPU thread:

```
Launch overhead 1 | Execution 1
Launch overhead 2 | Execution 2
Launch overhead 3 | Execution 3
...
```